# Project Big Data Yelp

**This notebook has been written using spark. Please open it with databricks**

In [2]:
spark.conf.set(
  "fs.azure.account.key.storagestudent.blob.core.windows.net", 
  "8ykAjWw0X+FVY8PnmVhnY5XyDdbpBLceLsFTbuTttdn2W76+sdAsSwYNQ9E+H8Oxlsay8/uCm8ayyWq73kB82Q=="
)

In [3]:
datasets = {
  dataset: spark.read.load( 
    "wasbs://default@storagestudent.blob.core.windows.net/datasets/S8-4/Exo/restaurant-data-with-consumer-ratings/{0}.csv".format(dataset), 
    format="csv",
    header="true"
  )
  for dataset in [
    "chefmozaccepts", 
    "rating_final", 
    "userpayment"
  ]
} 
ratings = datasets["rating_final"]
chef = datasets["chefmozaccepts"]
user_payment = datasets["userpayment"]

In [4]:
display(ratings)

userID placeID rating food_rating service_rating U1077 135085 2 2 2 U1077 135038 2 2 1 U1077 132825 2 2 2 U1077 135060 1 2 2 U1068 135104 1 1 2 U1068 132740 0 0 0 U1068 132663 1 1 1 U1068 132732 0 0 0 U1068 132630 1 1 1 U1067 132584 2 2 2 U1067 132733 1 1 1 U1067 132732 1 2 2 U1067 132630 1 0 1 U1067 135104 0 0 0 U1067 132560 1 0 0 U1103 132584 1 2 1 U1103 132732 0 0 2 U1103 132630 1 2 0 U1103 132613 2 2 2 U1103 132667 1 2 2 U1103 135104 1 2 0 U1103 132663 1 0 2 U1103 132733 2 2 2 U1107 132660 2 2 1 U1107 132584 2 2 2 U1107 132733 2 2 2 U1044 135088 2 2 2 U1044 132583 1 2 1 U1070 132608 2 2 1 U1070 132609 1 1 1 U1070 132613 1 1 0 U1031 132663 0 0 0 U1031 132665 0 0 0 U1031 132668 0 0 0 U1082 132630 1 1 1 U1023 132715 0 0 0 U1023 132717 0 0 0 U1082 132613 0 0 0 U1082 132594 0 0 0 U1082 132608 0 0 0 U1082 132609 0 0 0 U1082 132663 0 0 0 U1082 132560 0 0 0 U1060 132564 2 2 2 U1060 132717 2 2 2 U1060 132733 2 1 2 U1060 132740 1 2 1 U1015 135071 0 0 0 U1123 132733 1 2 1 U1123 132608 1 2 1 U1123 135104 2 2 2 U1123 132584 1 2 0 U1123 132667 1 2 1 U1123 132740 1 1 1 U1123 132613 1 2 1 U1123 132594 1 1 1 U1021 132740 2 2 1 U1026 132626 2 2 2 U1021 132668 2 2 2 U1021 132715 2 2 1 U1044 134987 0 0 0 U1083 135034 2 2 2 U1083 132723 1 1 1 U1083 135046 2 2 1 U1108 135075 2 2 2 U1108 132572 1 2 1 U1108 135074 2 2 1 U1108 135028 1 1 1 U1108 135085 1 2 1 U1108 135032 2 1 2 U1108 135086 1 1 0 U1108 132723 2 2 2 U1108 135046 1 2 1 U1108 135058 1 1 1 U1012 135001 1 1 2 U1012 134983 2 2 2 U1012 134996 2 2 2 U1093 135019 2 2 2 U1093 134996 2 2 2 U1093 135001 2 2 2 U1093 134976 2 2 1 U1093 134999 2 2 2 U1030 135018 2 2 2 U1093 135013 2 2 1 U1030 134986 2 2 2 U1012 134986 2 2 2 U1012 135018 2 2 2 U1030 135088 1 0 1 U1093 132767 2 2 2 U1030 135016 0 0 0 U1093 135011 1 1 0 U1030 132767 0 0 0 U1030 135019 0 0 0 U1030 135109 0 0 0 U1030 134996 0 0 0 U1066 135011 1 2 2 U1066 134999 1 1 1 U1127 134975 2 2 2 U1127 135021 2 1 1 U1127 134987 2 0 2 U1127 134996 2 2 2 U1017 132768 1 1 1 U1066 135019 2 1 0 U1017 134975 2 1 1 U1017 134983 0 0 1 U1066 132767 1 2 1 U1100 135021 2 1 0 U1100 134996 2 1 1 U1066 135013 1 2 2 U1100 135000 2 1 1 U1100 135001 2 2 2 U1100 132768 2 2 2 U1100 134976 2 2 1 U1133 134983 1 1 1 U1133 134986 2 2 2 U1133 132766 1 1 0 U1133 132767 0 0 0 U1133 135021 2 2 2 U1133 135001 2 2 2 U1133 134975 1 1 1 U1133 134992 1 2 1 U1133 135019 1 2 0 U1118 134987 0 0 0 U1072 135016 1 2 2 U1072 132773 2 1 2 U1072 135088 0 1 0 U1026 132561 1 1 1 U1026 132706 0 2 0 U1026 132654 0 0 0 U1026 132665 2 2 2 U1023 132740 0 0 0 U1023 132733 2 2 2 U1080 132668 1 1 1 U1080 132660 2 2 2 U1080 132733 1 1 1 U1080 132608 2 2 2 U1063 132660 2 2 2 U1063 132584 2 1 1 U1063 132609 1 0 1 U1063 135104 1 2 1 U1063 132667 1 2 1 U1080 132564 1 1 1 U1080 132715 1 1 1 U1082 132732 0 0 0 U1082 132733 0 0 0 U1074 135011 2 1 2 U1074 134987 2 2 2 U1074 134976 2 2 2 U1117 135018 1 1 1 U1117 132766 0 0 1 U1117 135088 1 1 2 U1117 135016 2 2 1 U1117 132773 1 2 2 U1117 135000 1 1 1 U1118 132583 0 0 0 U1118 132768 0 0 0 U1118 135021 2 0 0 U1118 134983 0 0 0 U1118 135001 0 0 0 U1118 134992 0 0 0 U1020 135088 1 2 0 U1020 132768 2 2 2 U1020 134996 1 1 1 U1020 135109 2 2 1 U1051 135088 1 1 1 U1051 135109 1 1 1 U1051 132768 1 2 1 U1055 135051 1 2 2 U1055 135075 2 2 2 U1055 132572 2 2 2 U1055 135028 2 2 2 U1055 135025 2 2 2 U1083 132572 2 2 2 U1083 135049 0 0 0 U1055 135106 2 2 2 U1083 135072 1 2 2 U1083 135074 2 2 2 U1083 132755 2 2 2 U1083 135044 1 2 2 U1055 135048 2 2 2 U1055 135047 2 2 2 U1055 132754 2 2 2 U1055 132825 2 2 2 U1053 135076 1 1 1 U1053 135032 2 2 1 U1053 132825 1 2 0 U1053 135063 2 1 2 U1035 135018 1 2 1 U1035 132773 2 1 2 U1053 135042 2 1 2 U1053 135053 2 2 0 U1053 132754 1 2 1 U1035 132583 1 0 2 U1053 135081 2 2 1 U1053 135025 2 1 2 U1053 135038 2 1 2 U1053 135060 2 2 1 U1053 135062 2 1 1 U1053 135047 2 2 2 U1064 135042 1 2 1 U1081 135069 1 0 1 U1064 135057 1 1 2 U1081 135076 1 2 2 U1081 135043 0 0 1 U1064 135069 0 0 1 U1081 135039 0 0 0 U1064 135060 1 0 1 U1064 135052 2 0 2

## Preprocessing

In [6]:
from pyspark.sql import functions as F
from pyspark.sql.types import *

In [7]:
length = F.udf(len, IntegerType())

In [8]:
display(ratings.select(length(F.col("userID"))).distinct())

len(userID) 5

all userID have the same length, I can use "substr" to remove the "U" on all the dataset 

In [10]:
ratings = ratings.select(F.col("userID").substr(2,5).cast(IntegerType()).alias("userID"),
                       F.col("placeID").cast(IntegerType()),
                       F.col("rating").cast(IntegerType())
                      )

In [11]:
display(ratings)

userID placeID rating 1077 135085 2 1077 135038 2 1077 132825 2 1077 135060 1 1068 135104 1 1068 132740 0 1068 132663 1 1068 132732 0 1068 132630 1 1067 132584 2 1067 132733 1 1067 132732 1 1067 132630 1 1067 135104 0 1067 132560 1 1103 132584 1 1103 132732 0 1103 132630 1 1103 132613 2 1103 132667 1 1103 135104 1 1103 132663 1 1103 132733 2 1107 132660 2 1107 132584 2 1107 132733 2 1044 135088 2 1044 132583 1 1070 132608 2 1070 132609 1 1070 132613 1 1031 132663 0 1031 132665 0 1031 132668 0 1082 132630 1 1023 132715 0 1023 132717 0 1082 132613 0 1082 132594 0 1082 132608 0 1082 132609 0 1082 132663 0 1082 132560 0 1060 132564 2 1060 132717 2 1060 132733 2 1060 132740 1 1015 135071 0 1123 132733 1 1123 132608 1 1123 135104 2 1123 132584 1 1123 132667 1 1123 132740 1 1123 132613 1 1123 132594 1 1021 132740 2 1026 132626 2 1021 132668 2 1021 132715 2 1044 134987 0 1083 135034 2 1083 132723 1 1083 135046 2 1108 135075 2 1108 132572 1 1108 135074 2 1108 135028 1 1108 135085 1 1108 135032 2 1108 135086 1 1108 132723 2 1108 135046 1 1108 135058 1 1012 135001 1 1012 134983 2 1012 134996 2 1093 135019 2 1093 134996 2 1093 135001 2 1093 134976 2 1093 134999 2 1030 135018 2 1093 135013 2 1030 134986 2 1012 134986 2 1012 135018 2 1030 135088 1 1093 132767 2 1030 135016 0 1093 135011 1 1030 132767 0 1030 135019 0 1030 135109 0 1030 134996 0 1066 135011 1 1066 134999 1 1127 134975 2 1127 135021 2 1127 134987 2 1127 134996 2 1017 132768 1 1066 135019 2 1017 134975 2 1017 134983 0 1066 132767 1 1100 135021 2 1100 134996 2 1066 135013 1 1100 135000 2 1100 135001 2 1100 132768 2 1100 134976 2 1133 134983 1 1133 134986 2 1133 132766 1 1133 132767 0 1133 135021 2 1133 135001 2 1133 134975 1 1133 134992 1 1133 135019 1 1118 134987 0 1072 135016 1 1072 132773 2 1072 135088 0 1026 132561 1 1026 132706 0 1026 132654 0 1026 132665 2 1023 132740 0 1023 132733 2 1080 132668 1 1080 132660 2 1080 132733 1 1080 132608 2 1063 132660 2 1063 132584 2 1063 132609 1 1063 135104 1 1063 132667 1 1080 132564 1 1080 132715 1 1082 132732 0 1082 132733 0 1074 135011 2 1074 134987 2 1074 134976 2 1117 135018 1 1117 132766 0 1117 135088 1 1117 135016 2 1117 132773 1 1117 135000 1 1118 132583 0 1118 132768 0 1118 135021 2 1118 134983 0 1118 135001 0 1118 134992 0 1020 135088 1 1020 132768 2 1020 134996 1 1020 135109 2 1051 135088 1 1051 135109 1 1051 132768 1 1055 135051 1 1055 135075 2 1055 132572 2 1055 135028 2 1055 135025 2 1083 132572 2 1083 135049 0 1055 135106 2 1083 135072 1 1083 135074 2 1083 132755 2 1083 135044 1 1055 135048 2 1055 135047 2 1055 132754 2 1055 132825 2 1053 135076 1 1053 135032 2 1053 132825 1 1053 135063 2 1035 135018 1 1035 132773 2 1053 135042 2 1053 135053 2 1053 132754 1 1035 132583 1 1053 135081 2 1053 135025 2 1053 135038 2 1053 135060 2 1053 135062 2 1053 135047 2 1064 135042 1 1081 135069 1 1064 135057 1 1081 135076 1 1081 135043 0 1064 135069 0 1081 135039 0 1064 135060 1 1064 135052 2 1081 132825 1 1081 135032 0 1081 135028 1 1081 135085 1 1081 135052 0 1081 135030 1 1081 135027 0 1001 132830 1 1001 132825 2 1014 135076 1 1088 132830 0 1088 135108 2 1014 135069 1 1014 135050 2 1088 135057 1 1014 135081 2 1014 135057 1 1014 135070 2 1088 135041 2 1088 135069 1 1088 135082 2 1014 135043 1 1014 135042 1 1088 135071 0 1088 135054 2 1088 135070 2 1014 135066 2 1088 135032 2 1088 135081 1 1088 135051 2 1014 132830 2 1014 135082 0 1056 135066 2 1056 135050 0 1056 135076 2 1056 135081 2 1056 135062 2 1056 135027 2 1056 135085 2 1056 135052 2 1056 132825 2 1056 135051 2 1134 135047 1 1134 135026 0 1134 135044 1 1134 132825 1 1134 135059 2 1134 135065 2 1134 135046 2 1134 135085 2 1134 135079 1 1134 135055 2 1134 132754 2 1134 132572 0 1134 135075 2 1134 135080 1 1134 135064 2 1134 135074 2 1106 135041 1 1126 135069 0 1106 135032 0 1126 135057 2 1126 132830 0 1106 135028 2 1126 135040 0 1126 135106 2 1126 135052 2 1126 135041 2 1126 135038 1 1126 135054 1 1126 135108 2 1044 134986 2 1124 135054 1 1044 132768 2 1124 135108 1 1124 135032 1 11

In [12]:
ratings.select("placeID").distinct().count()

Out[22]: 130

In [13]:
ratings.select("userID").distinct().count()

Out[23]: 138

This dataset contains 130 different restaurants and 138 different clients 

# Model Training

In [16]:
from pyspark.ml.recommendation import ALS
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.tuning import ParamGridBuilder, CrossValidator

In [17]:
train, test = ratings.randomSplit([0.7, 0.3], 100)

In [18]:
als = ALS(userCol = "userID", itemCol="placeID", ratingCol = "rating", coldStartStrategy="drop", nonnegative=True)
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
paramGrid = ParamGridBuilder()\
            .addGrid(als.rank, [1,5])\
            .addGrid(als.maxIter, [10,15]).build()

In [19]:
cv = CrossValidator(
  estimator = als,
  evaluator = evaluator,
  estimatorParamMaps = paramGrid,
  numFolds = 5
)

In [20]:
model = cv.fit(train)

# Evaluation

In [22]:
predictions = model.transform(test)
evaluator.evaluate(predictions)

Out[68]: 0.6802526241308007

In [23]:
display(predictions)

userID placeID rating prediction 1100 135000 2 1.5704908 1072 135000 0 1.5042318 1077 135027 0 1.0472511 1075 135066 2 0.7830303 1058 135066 1 1.6951612 1014 135066 2 1.5049499 1031 132663 0 0.0 1087 132663 1 0.32778248 1069 135108 0 0.52731025 1111 135108 2 1.3544235 1007 135108 1 1.1959543 1115 135108 1 1.2140629 1124 135071 1 0.1617671 1094 135071 0 0.0 1115 135071 2 1.034408 1092 135071 1 0.7642129 1048 132723 2 0.69113487 1096 132723 1 0.72814834 1003 132723 2 1.5418596 1083 132723 1 1.4301134 1095 132723 2 1.1852406 1137 135062 2 1.7373935 1132 135062 2 1.3323407 1073 135062 0 0.0 1022 135062 1 1.625203 1090 135062 2 1.4343879 1040 132773 1 0.81739 1112 132862 1 0.1477035 1090 132862 2 1.0832576 1009 132862 2 1.8255279 1055 135048 2 0.9551509 1006 135048 1 0.68838364 1130 132706 1 0.0 1028 132732 1 0.5084708 1043 132732 1 0.49106672 1058 132885 1 0.044358466 1082 132608 0 0.0 1043 132608 1 0.24539794 1016 132856 2 1.362486 1135 132856 0 0.0 1045 132856 1 0.6308767 1055 135025 2 1.9817828 1104 135025 2 0.5315131 1053 135025 2 0.7480026 1018 135025 1 0.1721323 1092 135025 2 0.23189552 1090 135025 2 1.3606799 1128 132925 0 0.0 1122 135030 2 1.8726515 1116 135030 1 1.6225739 1033 135030 2 1.6736058 1022 135040 0 0.3170848 1126 135040 0 0.51919466 1105 135040 0 0.0 1068 132740 0 0.37552232 1050 132740 0 0.0 1061 132921 1 1.0628955 1089 132921 1 0.9001169 1029 132921 1 -0.4943481 1136 132921 2 0.7969481 1027 132921 1 0.61245906 1033 132921 2 1.0558761 1138 132921 2 0.66601914 1133 135019 1 -0.36184222 1093 135019 2 0.90906054 1121 135019 2 1.3218666 1006 132572 1 1.0722201 1083 132572 2 0.84407914 1137 135049 2 0.40778357 1114 132951 0 0.100663066 1073 132951 0 0.0 1086 132951 1 1.7780521 1112 135026 1 0.35700274 1108 135046 1 1.0822775 1098 135046 1 0.62942374 1033 135046 2 0.9834494 1092 135046 0 0.18152916 1112 135046 0 -0.3435979 1058 135053 2 1.6269251 1097 135053 2 1.5567911 1055 135106 2 1.8671919 1002 135106 1 1.2461376 1101 135106 0 0.85611767 1029 132937 1 0.2991956 1003 132937 2 1.9835352 1027 132937 1 0.8702073 1042 134992 1 0.9196138 1056 135052 2 1.7994385 1073 135052 0 0.0 1049 135052 0 0.0 1078 135052 2 1.6118294 1054 135052 2 2.0574858 1076 135039 2 0.5813663 1001 135039 1 1.4479327 1081 135039 0 0.37990716 1024 135039 2 0.5241744 1106 135039 2 0.58769995 1034 135035 1 0.30292076 1095 135035 2 1.6231362 1124 135057 1 0.35493064 1075 135057 2 0.25967556 1094 135057 0 0.0 1007 135057 1 0.3076186 1126 135057 2 0.5540532 1120 135057 2 0.80509484 1026 132626 2 0.0 1039 132626 2 0.62809014 1030 132767 0 0.34757763 1093 132767 2 0.8924743 1066 132767 1 1.0913358 1024 135073 1 0.79937154 1119 134986 2 1.1106864 1005 132830 1 0.043188557 1014 132830 2 0.27711618 1088 135081 1 1.6362019 1124 135081 1 1.047686 1022 135081 0 1.6398413 1078 135081 1 0.94813263 1012 134983 2 0.9027175 1089 135079 0 1.0223675 1113 135079 1 0.4040129 1036 135079 2 1.4977908 1006 135079 1 1.3568748 1013 135079 1 1.8484806 1097 135079 2 1.1094978 1009 135079 1 0.8780564 1073 132954 0 0.0 1097 132954 1 1.1790638 1093 135011 1 0.47913027 1051 135109 1 0.21269125 1030 135109 0 1.1664345 1128 135060 0 0.0 1045 135060 2 0.3306249 1073 135060 0 0.0 1022 135060 1 0.71354747 1116 135060 0 1.469407 1098 135060 2 1.2849516 1053 135060 2 1.2894433 1054 135060 2 1.1753151 1013 135060 1 1.5868933 1055 132754 2 1.8636845 1053 132754 1 1.9281545 1082 132630 1 0.0 1028 132630 2 0.59235734 1055 135075 2 2.404488 1096 135075 1 1.6875454 1033 135075 2 1.7513281 1013 135075 2 1.2593457 1011 132715 1 0.08504397 1084 132861 2 1.4630164 1046 132861 1 1.1952213 1098 132861 2 0.71342635 1095 132861 1 1.249678 1114 132872 1 -0.11557999 1062 132872 0 0.0 1027 132872 1 0.5510998 1109 132872 1 1.6804454 1135 135045 0 0.0 1024 135045 2 1.2799078 1004 132955 2 2.0293205 1097 132955 2 1.3535155 1088 135082 2 -0.05155328 1115 135082 2 0.13678803 1018 135082 2 0.23727354 1103 135104 1 0.7451267 1082 132609 0 0.0 1050 132609 0 0.0 1070 132609 1 0.20986393 1063 132609 1 0.3893565

In [24]:
import pandas as pd
params = [{p.name: v for p, v in m.items()} for m in model.getEstimatorParamMaps()]
pd.DataFrame.from_dict([
    {model.getEvaluator().getMetricName(): metric, **ps} 
    for ps, metric in zip(params, model.avgMetrics)
]).sort_values("rmse")

maxIter 
 rank 
 rmse 
 
 
 
 
 1 
 15 
 1 
 0.748526 
 
 
 0 
 10 
 1 
 0.748696 
 
 
 2 
 10 
 5 
 0.837963 
 
 
 3 
 15 
 5 
 0.843146

# Recommendations

In [26]:
user_recommendations = model.bestModel.recommendForAllUsers(10)
display(user_recommendations)

userID recommendations 1088 List(List(132717, 1.9194726), List(134986, 1.6705197), List(132922, 1.6403724), List(132755, 1.6327808), List(132773, 1.619939), List(132937, 1.602175), List(132851, 1.5752245), List(132613, 1.5486337), List(135034, 1.5376678), List(135000, 1.5149316)) 1084 List(List(132717, 2.658769), List(134986, 2.3139303), List(132922, 2.2721715), List(132755, 2.2616558), List(132773, 2.243868), List(132937, 2.2192621), List(132851, 2.1819317), List(132613, 2.1450992), List(135034, 2.1299095), List(135000, 2.0984166)) 1025 List(List(132717, 2.3688438), List(134986, 2.061608), List(132922, 2.024403), List(132755, 2.015034), List(132773, 1.9991859), List(132937, 1.9772631), List(132851, 1.9440033), List(132613, 1.9111872), List(135034, 1.8976539), List(135000, 1.869595)) 1127 List(List(132717, 3.2231047), List(134986, 2.8050725), List(132922, 2.7544503), List(132755, 2.7417026), List(132773, 2.7201393), List(132937, 2.6903107), List(132851, 2.6450565), List(132613, 2.6004062), List(135034, 2.5819926), List(135000, 2.543815)) 1133 List(List(132717, 2.053721), List(134986, 1.7873563), List(132922, 1.7551005), List(132755, 1.7469778), List(132773, 1.7332379), List(132937, 1.7142315), List(132851, 1.6853962), List(132613, 1.6569455), List(135034, 1.6452125), List(135000, 1.6208863)) 1005 List(List(132717, 2.5342004), List(134986, 2.2055182), List(132922, 2.165716), List(132755, 2.155693), List(132773, 2.1387386), List(132937, 2.1152856), List(132851, 2.079704), List(132613, 2.0445971), List(135034, 2.0301192), List(135000, 2.0001018)) 1016 List(List(132717, 2.8693225), List(134986, 2.4971755), List(132922, 2.4521098), List(132755, 2.4407616), List(132773, 2.421565), List(132937, 2.3950105), List(132851, 2.3547237), List(132613, 2.3149743), List(135034, 2.2985818), List(135000, 2.2645948)) 1068 List(List(132717, 1.1728821), List(134986, 1.0207609), List(132922, 1.0023396), List(132755, 0.9977007), List(132773, 0.98985386), List(132937, 0.97899926), List(132851, 0.9625314), List(132613, 0.94628316), List(135034, 0.93958247), List(135000, 0.92568976)) 1031 List(List(132560, 0.0), List(132630, 0.0), List(132660, 0.0), List(132740, 0.0), List(132830, 0.0), List(132870, 0.0), List(135000, 0.0), List(135030, 0.0), List(135040, 0.0), List(135050, 0.0)) 1051 List(List(132717, 1.2751657), List(134986, 1.1097785), List(132922, 1.0897508), List(132755, 1.0847074), List(132773, 1.0761762), List(132937, 1.064375), List(132851, 1.046471), List(132613, 1.0288059), List(135034, 1.0215207), List(135000, 1.0064164)) 1030 List(List(132717, 1.1247872), List(134986, 0.97890383), List(132922, 0.9612379), List(132755, 0.9567893), List(132773, 0.94926417), List(132937, 0.93885475), List(132851, 0.9230621), List(132613, 0.9074802), List(135034, 0.90105426), List(135000, 0.8877312)) 1034 List(List(132717, 0.57401824), List(134986, 0.49956888), List(132922, 0.49055332), List(132755, 0.48828304), List(132773, 0.4844427), List(132937, 0.47913042), List(132851, 0.4710709), List(132613, 0.46311888), List(135034, 0.4598395), List(135000, 0.45304027)) 1064 List(List(132717, 1.8309624), List(134986, 1.5934892), List(132922, 1.564732), List(132755, 1.5574905), List(132773, 1.5452408), List(132937, 1.528296), List(132851, 1.5025883), List(132613, 1.4772235), List(135034, 1.4667633), List(135000, 1.4450755)) 1019 List(List(132560, 0.0), List(132630, 0.0), List(132660, 0.0), List(132740, 0.0), List(132830, 0.0), List(132870, 0.0), List(135000, 0.0), List(135030, 0.0), List(135040, 0.0), List(135050, 0.0)) 1135 List(List(132560, 0.0), List(132630, 0.0), List(132660, 0.0), List(132740, 0.0), List(132830, 0.0), List(132870, 0.0), List(135000, 0.0), List(135030, 0.0), List(135040, 0.0), List(135050, 0.0)) 1085 List(List(132717, 2.3057463), List(134986, 2.006694), List(132922, 1.9704801), List(132755, 1.9613607), List(132773, 1.9459347), List(132937, 1.9245958), List(132851, 1.8922219), List(132613, 1.8602799), List(135034, 1.8471072), List(135000, 1.8197957)) 

In [27]:
item_recommendations = model.bestModel.recommendForAllItems(10)
display(item_recommendations)

placeID recommendations 135000 List(List(1021, 3.1574771), List(1055, 2.631266), List(1020, 2.5615323), List(1127, 2.543815), List(1052, 2.5283103), List(1078, 2.4990141), List(1102, 2.3995512), List(1110, 2.3615167), List(1071, 2.3075306), List(1054, 2.3022351)) 135027 List(List(1021, 1.9296571), List(1055, 1.6080691), List(1020, 1.5654521), List(1127, 1.5546243), List(1052, 1.5451487), List(1078, 1.5272448), List(1102, 1.466459), List(1110, 1.4432147), List(1071, 1.4102218), List(1054, 1.4069854)) 135066 List(List(1021, 2.690701), List(1055, 2.2422807), List(1020, 2.1828558), List(1127, 2.1677575), List(1052, 2.154545), List(1078, 2.1295798), List(1102, 2.0448205), List(1110, 2.0124087), List(1071, 1.9664037), List(1054, 1.9618909)) 132663 List(List(1021, 1.4287605), List(1055, 1.1906496), List(1020, 1.159095), List(1127, 1.151078), List(1052, 1.1440622), List(1078, 1.1308056), List(1102, 1.0857985), List(1110, 1.0685879), List(1071, 1.0441592), List(1054, 1.0417628)) 135108 List(List(1021, 3.1500924), List(1055, 2.625112), List(1020, 2.5555413), List(1127, 2.5378654), List(1052, 2.522397), List(1078, 2.4931693), List(1102, 2.393939), List(1110, 2.3559935), List(1071, 2.3021338), List(1054, 2.2968504)) 135071 List(List(1021, 2.1168242), List(1055, 1.7640436), List(1020, 1.7172929), List(1127, 1.7054149), List(1052, 1.6950203), List(1078, 1.6753798), List(1102, 1.6086981), List(1110, 1.5831991), List(1071, 1.5470061), List(1054, 1.5434558)) 132723 List(List(1021, 2.609916), List(1055, 2.174959), List(1020, 2.1173182), List(1127, 2.1026733), List(1052, 2.0898576), List(1078, 2.0656419), List(1102, 1.9834274), List(1110, 1.9519887), List(1071, 1.9073648), List(1054, 1.9029876)) 135062 List(List(1021, 2.5019517), List(1055, 2.0849876), List(1020, 2.0297313), List(1127, 2.0156922), List(1052, 2.0034065), List(1078, 1.9801925), List(1102, 1.901379), List(1110, 1.8712409), List(1071, 1.828463), List(1054, 1.8242668)) 132862 List(List(1021, 2.543925), List(1055, 2.1199658), List(1020, 2.0637825), List(1127, 2.0495079), List(1052, 2.0370162), List(1078, 2.0134127), List(1102, 1.933277), List(1110, 1.9026333), List(1071, 1.8591378), List(1054, 1.8548712)) 132773 List(List(1021, 3.3763375), List(1055, 2.813652), List(1020, 2.7390847), List(1127, 2.7201393), List(1052, 2.7035599), List(1078, 2.672233), List(1102, 2.5658758), List(1110, 2.525205), List(1071, 2.467477), List(1054, 2.4618142)) 135048 List(List(1021, 1.5660675), List(1055, 1.3050736), List(1020, 1.2704866), List(1127, 1.261699), List(1052, 1.2540089), List(1078, 1.2394783), List(1102, 1.190146), List(1110, 1.1712813), List(1071, 1.144505), List(1054, 1.1418785)) 132732 List(List(1021, 1.6892478), List(1055, 1.4077253), List(1020, 1.3704178), List(1127, 1.360939), List(1052, 1.3526441), List(1078, 1.3369706), List(1102, 1.2837579), List(1110, 1.2634095), List(1071, 1.2345271), List(1054, 1.231694)) 132706 List(List(1021, 2.8063009), List(1055, 2.3386152), List(1020, 2.2766373), List(1127, 2.2608905), List(1052, 2.2471101), List(1078, 2.2210724), List(1102, 2.1326716), List(1110, 2.0988674), List(1071, 2.050886), List(1054, 2.046179)) 132885 List(List(1021, 0.87421525), List(1055, 0.7285224), List(1020, 0.7092151), List(1127, 0.7043097), List(1052, 0.7000169), List(1078, 0.6919056), List(1102, 0.66436714), List(1110, 0.6538365), List(1071, 0.6388893), List(1054, 0.6374231)) 135025 List(List(1021, 2.428424), List(1055, 2.0237136), List(1020, 1.9700811), List(1127, 1.9564546), List(1052, 1.94453), List(1078, 1.9219983), List(1102, 1.845501), List(1110, 1.8162485), List(1071, 1.7747278), List(1054, 1.7706549)) 132856 List(List(1021, 1.7052217), List(1055, 1.4210371), List(1020, 1.3833767), List(1127, 1.3738083), List(1052, 1.3654349), List(1078, 1.3496133), List(1102, 1.2958974), List(1110, 1.2753565), List(1071, 1.246201), List(1054, 1.2433411)) 132608 List(List(1021, 3.0056915), List(1055, 2.5047762), List(1020, 2.4383948), List(1127, 2.421529), List(1052, 2.4067698), List(107

# Getting in precision ...

In [29]:
ratings_predictions = predictions.select(
  F.concat(F.lit("U"), F.col("userID")).alias("userID"),
  F.col("placeID"),
  F.col("rating"),
  F.col("prediction")
)

In [30]:
display(ratings_predictions)

userID placeID rating prediction U1100 135000 2 1.5704908 U1072 135000 0 1.5042318 U1077 135027 0 1.0472511 U1075 135066 2 0.7830303 U1058 135066 1 1.6951612 U1014 135066 2 1.5049499 U1031 132663 0 0.0 U1087 132663 1 0.32778248 U1069 135108 0 0.52731025 U1111 135108 2 1.3544235 U1007 135108 1 1.1959543 U1115 135108 1 1.2140629 U1124 135071 1 0.1617671 U1094 135071 0 0.0 U1115 135071 2 1.034408 U1092 135071 1 0.7642129 U1048 132723 2 0.69113487 U1096 132723 1 0.72814834 U1003 132723 2 1.5418596 U1083 132723 1 1.4301134 U1095 132723 2 1.1852406 U1137 135062 2 1.7373935 U1132 135062 2 1.3323407 U1073 135062 0 0.0 U1022 135062 1 1.625203 U1090 135062 2 1.4343879 U1040 132773 1 0.81739 U1112 132862 1 0.1477035 U1090 132862 2 1.0832576 U1009 132862 2 1.8255279 U1055 135048 2 0.9551509 U1006 135048 1 0.68838364 U1130 132706 1 0.0 U1028 132732 1 0.5084708 U1043 132732 1 0.49106672 U1058 132885 1 0.044358466 U1082 132608 0 0.0 U1043 132608 1 0.24539794 U1016 132856 2 1.362486 U1135 132856 0 0.0 U1045 132856 1 0.6308767 U1055 135025 2 1.9817828 U1104 135025 2 0.5315131 U1053 135025 2 0.7480026 U1018 135025 1 0.1721323 U1092 135025 2 0.23189552 U1090 135025 2 1.3606799 U1128 132925 0 0.0 U1122 135030 2 1.8726515 U1116 135030 1 1.6225739 U1033 135030 2 1.6736058 U1022 135040 0 0.3170848 U1126 135040 0 0.51919466 U1105 135040 0 0.0 U1068 132740 0 0.37552232 U1050 132740 0 0.0 U1061 132921 1 1.0628955 U1089 132921 1 0.9001169 U1029 132921 1 -0.4943481 U1136 132921 2 0.7969481 U1027 132921 1 0.61245906 U1033 132921 2 1.0558761 U1138 132921 2 0.66601914 U1133 135019 1 -0.36184222 U1093 135019 2 0.90906054 U1121 135019 2 1.3218666 U1006 132572 1 1.0722201 U1083 132572 2 0.84407914 U1137 135049 2 0.40778357 U1114 132951 0 0.100663066 U1073 132951 0 0.0 U1086 132951 1 1.7780521 U1112 135026 1 0.35700274 U1108 135046 1 1.0822775 U1098 135046 1 0.62942374 U1033 135046 2 0.9834494 U1092 135046 0 0.18152916 U1112 135046 0 -0.3435979 U1058 135053 2 1.6269251 U1097 135053 2 1.5567911 U1055 135106 2 1.8671919 U1002 135106 1 1.2461376 U1101 135106 0 0.85611767 U1029 132937 1 0.2991956 U1003 132937 2 1.9835352 U1027 132937 1 0.8702073 U1042 134992 1 0.9196138 U1056 135052 2 1.7994385 U1073 135052 0 0.0 U1049 135052 0 0.0 U1078 135052 2 1.6118294 U1054 135052 2 2.0574858 U1076 135039 2 0.5813663 U1001 135039 1 1.4479327 U1081 135039 0 0.37990716 U1024 135039 2 0.5241744 U1106 135039 2 0.58769995 U1034 135035 1 0.30292076 U1095 135035 2 1.6231362 U1124 135057 1 0.35493064 U1075 135057 2 0.25967556 U1094 135057 0 0.0 U1007 135057 1 0.3076186 U1126 135057 2 0.5540532 U1120 135057 2 0.80509484 U1026 132626 2 0.0 U1039 132626 2 0.62809014 U1030 132767 0 0.34757763 U1093 132767 2 0.8924743 U1066 132767 1 1.0913358 U1024 135073 1 0.79937154 U1119 134986 2 1.1106864 U1005 132830 1 0.043188557 U1014 132830 2 0.27711618 U1088 135081 1 1.6362019 U1124 135081 1 1.047686 U1022 135081 0 1.6398413 U1078 135081 1 0.94813263 U1012 134983 2 0.9027175 U1089 135079 0 1.0223675 U1113 135079 1 0.4040129 U1036 135079 2 1.4977908 U1006 135079 1 1.3568748 U1013 135079 1 1.8484806 U1097 135079 2 1.1094978 U1009 135079 1 0.8780564 U1073 132954 0 0.0 U1097 132954 1 1.1790638 U1093 135011 1 0.47913027 U1051 135109 1 0.21269125 U1030 135109 0 1.1664345 U1128 135060 0 0.0 U1045 135060 2 0.3306249 U1073 135060 0 0.0 U1022 135060 1 0.71354747 U1116 135060 0 1.469407 U1098 135060 2 1.2849516 U1053 135060 2 1.2894433 U1054 135060 2 1.1753151 U1013 135060 1 1.5868933 U1055 132754 2 1.8636845 U1053 132754 1 1.9281545 U1082 132630 1 0.0 U1028 132630 2 0.59235734 U1055 135075 2 2.404488 U1096 135075 1 1.6875454 U1033 135075 2 1.7513281 U1013 135075 2 1.2593457 U1011 132715 1 0.08504397 U1084 132861 2 1.4630164 U1046 132861 1 1.1952213 U1098 132861 2 0.71342635 U1095 132861 1 1.249678 U1114 132872 1 -0.11557999 U1062 132872 0 0.0 U1027 132872 1 0.5510998 U1109 132872 1 1.6804454 U1135 135045 0 0.0 U1024 135045 2 1.2799078 U1004 132955 2 2.0293205 U1097 132955 2 1.3535155 U1088 135082 2 -0.05155

In [31]:
display(chef)

placeID Rpayment 135110 cash 135110 VISA 135110 MasterCard-Eurocard 135110 American_Express 135110 bank_debit_cards 135109 cash 135107 cash 135107 VISA 135107 MasterCard-Eurocard 135107 American_Express 135107 bank_debit_cards 135106 cash 135106 VISA 135106 MasterCard-Eurocard 135105 cash 135104 cash 135104 VISA 135104 MasterCard-Eurocard 135103 cash 135103 VISA 135103 MasterCard-Eurocard 135102 cash 135102 VISA 135102 MasterCard-Eurocard 135102 bank_debit_cards 135101 cash 135100 cash 135099 cash 135099 VISA 135099 MasterCard-Eurocard 135098 cash 135098 VISA 135098 MasterCard-Eurocard 135098 bank_debit_cards 135097 cash 135097 VISA 135097 MasterCard-Eurocard 135097 bank_debit_cards 135094 cash 135093 cash 135093 bank_debit_cards 135092 cash 135091 cash 135091 VISA 135091 MasterCard-Eurocard 135090 cash 135089 cash 135089 VISA 135089 MasterCard-Eurocard 135089 bank_debit_cards 135088 cash 135086 cash 135086 VISA 135086 MasterCard-Eurocard 135085 cash 135084 cash 135084 VISA 135084 MasterCard-Eurocard 135082 cash 135081 cash 135081 VISA 135081 MasterCard-Eurocard 135080 cash 135080 VISA 135080 MasterCard-Eurocard 135080 American_Express 135079 cash 135079 American_Express 135078 cash 135078 VISA 135078 MasterCard-Eurocard 135077 cash 135076 cash 135076 VISA 135076 MasterCard-Eurocard 135076 American_Express 135075 cash 135074 cash 135074 VISA 135074 MasterCard-Eurocard 135073 cash 135073 VISA 135073 MasterCard-Eurocard 135072 cash 135072 VISA 135072 MasterCard-Eurocard 135072 American_Express 135071 cash 135071 VISA 135071 MasterCard-Eurocard 135071 American_Express 135070 cash 135069 cash 135068 cash 135068 VISA 135068 MasterCard-Eurocard 135067 cash 135067 VISA 135067 MasterCard-Eurocard 135066 cash 135066 VISA 135066 MasterCard-Eurocard 135065 cash 135065 VISA 135065 MasterCard-Eurocard 135065 American_Express 135064 cash 135064 VISA 135064 MasterCard-Eurocard 135063 cash 135062 cash 135061 cash 135061 VISA 135061 MasterCard-Eurocard 135061 American_Express 135060 cash 135059 cash 135058 cash 135058 VISA 135058 MasterCard-Eurocard 135057 cash 135057 VISA 135057 MasterCard-Eurocard 135057 American_Express 135056 cash 135055 cash 135055 VISA 135055 MasterCard-Eurocard 135055 American_Express 135054 cash 135054 VISA 135054 MasterCard-Eurocard 135053 cash 135053 VISA 135053 MasterCard-Eurocard 135053 American_Express 135052 cash 135052 VISA 135052 MasterCard-Eurocard 135051 cash 135050 cash 135050 VISA 135050 MasterCard-Eurocard 135050 American_Express 135049 cash 135048 cash 135048 VISA 135048 MasterCard-Eurocard 135048 American_Express 135047 cash 135047 VISA 135047 MasterCard-Eurocard 135046 cash 135046 VISA 135046 MasterCard-Eurocard 135045 cash 135045 VISA 135045 MasterCard-Eurocard 135045 American_Express 135044 cash 135043 cash 135042 cash 135041 cash 135040 cash 135040 VISA 135040 MasterCard-Eurocard 135039 cash 135038 cash 135037 cash 135037 VISA 135037 MasterCard-Eurocard 135036 cash 135035 cash 135035 VISA 135035 MasterCard-Eurocard 135035 American_Express 135034 cash 135034 VISA 135034 MasterCard-Eurocard 135034 American_Express 135033 cash 135033 VISA 135033 MasterCard-Eurocard 135033 American_Express 135032 cash 135032 VISA 135032 MasterCard-Eurocard 135031 cash 135031 VISA 135031 MasterCard-Eurocard 135030 cash 135029 cash 135029 VISA 135029 MasterCard-Eurocard 135029 American_Express 135028 cash 135028 VISA 135028 MasterCard-Eurocard 135028 American_Express 135027 cash 135027 VISA 135027 MasterCard-Eurocard 135027 American_Express 135025 cash 135025 VISA 135025 MasterCard-Eurocard 135025 American_Express 135024 cash 135024 VISA 135024 MasterCard-Eurocard 135024 American_Express 135023 cash 135022 cash 135021 cash 135018 cash 135018 VISA 135018 MasterCard-Eurocard 135018 bank_debit_cards 135017 cash 135016 cash 135015 cash 135014 cash 135013 cash 135012 cash 135011 cash 135010 cash 135009 cash 135008 cash 135008 VISA 135008 MasterCard-Eurocard 135008 American_Express 135007 VISA 135007 MasterCard-Eurocard 135007 A

In [32]:
display(user_payment)

userID Upayment U1001 cash U1002 cash U1003 cash U1004 cash U1004 bank_debit_cards U1005 cash U1006 cash U1007 cash U1008 cash U1009 cash U1010 cash U1011 cash U1012 cash U1012 bank_debit_cards U1013 MasterCard-Eurocard U1013 cash U1014 cash U1015 cash U1016 cash U1017 cash U1017 VISA U1018 cash U1019 cash U1020 cash U1020 bank_debit_cards U1021 cash U1022 cash U1023 cash U1026 cash U1026 bank_debit_cards U1027 cash U1028 cash U1029 cash U1030 cash U1031 cash U1031 American_Express U1032 cash U1033 cash U1034 cash U1035 cash U1035 VISA U1036 cash U1037 cash U1038 cash U1039 cash U1040 cash U1041 bank_debit_cards U1041 cash U1041 American_Express U1041 VISA U1042 bank_debit_cards U1043 cash U1044 VISA U1044 bank_debit_cards U1044 cash U1044 MasterCard-Eurocard U1045 cash U1046 cash U1047 cash U1048 cash U1049 cash U1050 cash U1051 VISA U1051 cash U1052 cash U1053 cash U1054 cash U1055 cash U1056 cash U1057 cash U1057 VISA U1058 cash U1059 cash U1060 cash U1061 cash U1061 bank_debit_cards U1062 VISA U1062 cash U1063 cash U1064 cash U1065 cash U1066 cash U1066 bank_debit_cards U1067 cash U1068 cash U1069 MasterCard-Eurocard U1069 cash U1070 cash U1071 cash U1072 bank_debit_cards U1072 cash U1073 cash U1074 bank_debit_cards U1074 cash U1075 cash U1076 VISA U1076 cash U1076 bank_debit_cards U1077 VISA U1077 cash U1077 bank_debit_cards U1078 VISA U1078 bank_debit_cards U1078 cash U1079 cash U1080 cash U1081 cash U1082 cash U1083 cash U1084 cash U1085 cash U1086 VISA U1086 bank_debit_cards U1086 cash U1087 cash U1089 cash U1090 cash U1091 cash U1092 cash U1093 American_Express U1093 VISA U1094 cash U1095 cash U1096 cash U1097 cash U1098 cash U1099 cash U1099 VISA U1100 cash U1101 cash U1102 cash U1103 cash U1103 bank_debit_cards U1104 VISA U1104 cash U1105 cash U1106 bank_debit_cards U1106 cash U1107 bank_debit_cards U1107 cash U1108 VISA U1108 cash U1108 MasterCard-Eurocard U1109 cash U1110 cash U1111 cash U1112 cash U1113 VISA U1113 cash U1114 cash U1115 cash U1116 cash U1116 bank_debit_cards U1116 VISA U1117 bank_debit_cards U1117 cash U1118 cash U1119 cash U1120 cash U1121 cash U1121 bank_debit_cards U1123 cash U1124 cash U1125 cash U1126 cash U1127 cash U1128 cash U1129 cash U1131 cash U1132 cash U1133 bank_debit_cards U1133 cash U1134 cash U1135 cash U1136 cash U1137 cash U1138 cash

In [33]:
payment_matching = user_payment.join(chef, user_payment["Upayment"]==chef["Rpayment"], how="inner")
payment_matching = payment_matching.select("userID", "placeID").dropDuplicates()

In [34]:
dataset = ratings_predictions.join(payment_matching, ["userID","placeID"], how="inner")
display(dataset)

userID placeID rating prediction U1100 135000 2 1.5704908 U1072 135000 0 1.5042318 U1077 135027 0 1.0472511 U1075 135066 2 0.7830303 U1058 135066 1 1.6951612 U1014 135066 2 1.5049499 U1124 135071 1 0.1617671 U1094 135071 0 0.0 U1115 135071 2 1.034408 U1092 135071 1 0.7642129 U1048 132723 2 0.69113487 U1096 132723 1 0.72814834 U1003 132723 2 1.5418596 U1083 132723 1 1.4301134 U1095 132723 2 1.1852406 U1137 135062 2 1.7373935 U1132 135062 2 1.3323407 U1073 135062 0 0.0 U1022 135062 1 1.625203 U1090 135062 2 1.4343879 U1040 132773 1 0.81739 U1112 132862 1 0.1477035 U1090 132862 2 1.0832576 U1009 132862 2 1.8255279 U1055 135048 2 0.9551509 U1006 135048 1 0.68838364 U1028 132732 1 0.5084708 U1043 132732 1 0.49106672 U1058 132885 1 0.044358466 U1082 132608 0 0.0 U1043 132608 1 0.24539794 U1016 132856 2 1.362486 U1135 132856 0 0.0 U1045 132856 1 0.6308767 U1055 135025 2 1.9817828 U1104 135025 2 0.5315131 U1053 135025 2 0.7480026 U1018 135025 1 0.1721323 U1092 135025 2 0.23189552 U1090 135025 2 1.3606799 U1128 132925 0 0.0 U1116 135030 1 1.6225739 U1033 135030 2 1.6736058 U1022 135040 0 0.3170848 U1126 135040 0 0.51919466 U1105 135040 0 0.0 U1068 132740 0 0.37552232 U1050 132740 0 0.0 U1061 132921 1 1.0628955 U1089 132921 1 0.9001169 U1029 132921 1 -0.4943481 U1136 132921 2 0.7969481 U1027 132921 1 0.61245906 U1033 132921 2 1.0558761 U1138 132921 2 0.66601914 U1006 132572 1 1.0722201 U1083 132572 2 0.84407914 U1137 135049 2 0.40778357 U1114 132951 0 0.100663066 U1073 132951 0 0.0 U1086 132951 1 1.7780521 U1108 135046 1 1.0822775 U1098 135046 1 0.62942374 U1033 135046 2 0.9834494 U1092 135046 0 0.18152916 U1112 135046 0 -0.3435979 U1058 135053 2 1.6269251 U1097 135053 2 1.5567911 U1055 135106 2 1.8671919 U1002 135106 1 1.2461376 U1101 135106 0 0.85611767 U1029 132937 1 0.2991956 U1003 132937 2 1.9835352 U1027 132937 1 0.8702073 U1056 135052 2 1.7994385 U1073 135052 0 0.0 U1049 135052 0 0.0 U1078 135052 2 1.6118294 U1054 135052 2 2.0574858 U1076 135039 2 0.5813663 U1001 135039 1 1.4479327 U1081 135039 0 0.37990716 U1106 135039 2 0.58769995 U1034 135035 1 0.30292076 U1095 135035 2 1.6231362 U1124 135057 1 0.35493064 U1075 135057 2 0.25967556 U1094 135057 0 0.0 U1007 135057 1 0.3076186 U1126 135057 2 0.5540532 U1120 135057 2 0.80509484 U1026 132626 2 0.0 U1039 132626 2 0.62809014 U1030 132767 0 0.34757763 U1093 132767 2 0.8924743 U1066 132767 1 1.0913358 U1124 135081 1 1.047686 U1022 135081 0 1.6398413 U1078 135081 1 0.94813263 U1012 134983 2 0.9027175 U1089 135079 0 1.0223675 U1113 135079 1 0.4040129 U1036 135079 2 1.4977908 U1006 135079 1 1.3568748 U1013 135079 1 1.8484806 U1097 135079 2 1.1094978 U1009 135079 1 0.8780564 U1073 132954 0 0.0 U1097 132954 1 1.1790638 U1051 135109 1 0.21269125 U1030 135109 0 1.1664345 U1128 135060 0 0.0 U1045 135060 2 0.3306249 U1073 135060 0 0.0 U1022 135060 1 0.71354747 U1116 135060 0 1.469407 U1098 135060 2 1.2849516 U1053 135060 2 1.2894433 U1054 135060 2 1.1753151 U1013 135060 1 1.5868933 U1055 132754 2 1.8636845 U1053 132754 1 1.9281545 U1055 135075 2 2.404488 U1096 135075 1 1.6875454 U1033 135075 2 1.7513281 U1013 135075 2 1.2593457 U1011 132715 1 0.08504397 U1084 132861 2 1.4630164 U1046 132861 1 1.1952213 U1098 132861 2 0.71342635 U1095 132861 1 1.249678 U1114 132872 1 -0.11557999 U1062 132872 0 0.0 U1027 132872 1 0.5510998 U1109 132872 1 1.6804454 U1135 135045 0 0.0 U1004 132955 2 2.0293205 U1097 132955 2 1.3535155 U1115 135082 2 0.13678803 U1018 135082 2 0.23727354 U1103 135104 1 0.7451267 U1082 132609 0 0.0 U1050 132609 0 0.0 U1070 132609 1 0.20986393 U1063 132609 1 0.3893565 U1126 135054 1 0.6613918 U1008 132869 1 0.8063797 U1015 132869 2 -0.36932832 U1131 132869 1 0.5259348 U1085 135059 2 0.88246816 U1137 135059 2 2.4445972 U1002 135059 1 0.5269936 U1004 132958 2 1.9603679 U1089 135028 1 0.7723479 U1132 135028 2 1.3217709 U1059 135028 1 1.1281946 U1081 135028 1 0.42973453 U1071 135028 2 1.9419383 U1033 135028 2 1.2892061 U1106 135028 2 1.3087524 U1004 135028 2 1.6987983 U1109 135028 2 1.52345

In [35]:
dataset.count()

Out[59]: 306

In [36]:
test_2 = dataset.select(F.col("userID").substr(2,5).cast(IntegerType()).alias("userID"),
                       F.col("placeID").cast(IntegerType()),
                       F.col("rating").cast(IntegerType())
                       )

In [37]:
predictions_2 = model.transform(test_2)
evaluator.evaluate(predictions_2)

Out[72]: 0.6766377868304377

In [38]:
display(predictions_2)

userID placeID rating prediction 1100 135000 2 2.1271644 1072 135000 0 1.5427417 1077 135027 0 1.2429208 1075 135066 2 1.3838029 1058 135066 1 1.891922 1014 135066 2 1.436259 1124 135071 1 0.815222 1094 135071 0 0.0 1115 135071 2 1.0773803 1092 135071 1 0.4832484 1048 132723 2 1.4554642 1096 132723 1 1.5241543 1003 132723 2 1.5700753 1083 132723 1 1.3671134 1095 132723 2 1.5477859 1137 135062 2 1.6113443 1132 135062 2 1.4168313 1073 135062 0 0.0 1022 135062 1 1.5473262 1090 135062 2 1.5545644 1040 132773 1 1.7108476 1112 132862 1 0.3166064 1090 132862 2 1.5806441 1009 132862 2 1.4877393 1055 135048 2 1.3050736 1006 135048 1 0.74235135 1028 132732 1 0.63748765 1043 132732 1 0.5812552 1058 132885 1 0.61469 1082 132608 0 0.0 1043 132608 1 1.0342318 1016 132856 2 1.2230132 1135 132856 0 0.0 1045 132856 1 0.8472481 1055 135025 2 2.0237136 1104 135025 2 1.1546063 1053 135025 2 1.7253983 1018 135025 1 0.6531835 1092 135025 2 0.55438334 1090 135025 2 1.5088786 1128 132925 0 0.0 1116 135030 1 1.655504 1033 135030 2 1.6525886 1022 135040 0 1.0251113 1126 135040 0 0.91323376 1105 135040 0 0.0 1068 132740 0 0.5018673 1050 132740 0 0.0 1061 132921 1 1.6548915 1089 132921 1 0.97932136 1029 132921 1 0.73188233 1136 132921 2 1.7179248 1027 132921 1 0.8094937 1033 132921 2 1.5564635 1138 132921 2 1.0003767 1006 132572 1 0.8476044 1083 132572 2 0.93663883 1137 135049 2 1.0735846 1114 132951 0 0.15683062 1073 132951 0 0.0 1086 132951 1 1.5615156 1108 135046 1 1.8307391 1098 135046 1 1.4841475 1033 135046 2 1.7080815 1092 135046 0 0.68208104 1112 135046 0 0.37184823 1058 135053 2 1.8030496 1097 135053 2 1.6882174 1055 135106 2 1.9628934 1002 135106 1 1.1996908 1101 135106 0 1.1576821 1029 132937 1 0.8976722 1003 132937 2 2.0088665 1027 132937 1 0.9928645 1056 135052 2 1.7738318 1073 135052 0 0.0 1049 135052 0 0.0 1078 135052 2 2.1187525 1054 135052 2 1.9519161 1076 135039 2 1.2809628 1001 135039 1 1.3027532 1081 135039 0 0.4383114 1106 135039 2 1.1628364 1034 135035 1 0.3809902 1095 135035 2 1.5747136 1124 135057 1 0.9544151 1075 135057 2 1.2745444 1094 135057 0 0.0 1007 135057 1 0.8570977 1126 135057 2 1.3654006 1120 135057 2 0.9366498 1026 132626 2 0.0 1039 132626 2 0.63305324 1030 132767 0 0.61102366 1093 132767 2 1.4754518 1066 132767 1 1.0505869 1124 135081 1 1.008022 1022 135081 0 1.6187577 1078 135081 1 2.0716069 1012 134983 2 1.3144791 1089 135079 0 1.0381075 1113 135079 1 1.165437 1036 135079 2 2.0836298 1006 135079 1 1.3680336 1013 135079 1 1.8069205 1097 135079 2 1.9000115 1009 135079 1 1.6877972 1073 132954 0 0.0 1097 132954 1 1.7611402 1051 135109 1 0.63836896 1030 135109 0 0.563087 1128 135060 0 0.0 1045 135060 2 0.98503697 1073 135060 0 0.0 1022 135060 1 1.2260996 1116 135060 0 1.135394 1098 135060 2 0.9848035 1053 135060 2 1.4085999 1054 135060 2 1.445547 1013 135060 1 1.2412637 1055 132754 2 2.303529 1053 132754 1 1.9639663 1055 135075 2 2.4921217 1096 135075 1 1.7464137 1033 135075 2 1.7096335 1013 135075 2 1.8723456 1011 132715 1 0.74654865 1084 132861 2 1.7394602 1046 132861 1 1.2619736 1098 132861 2 1.3001391 1095 132861 1 1.552199 1114 132872 1 0.11080006 1062 132872 0 0.0 1027 132872 1 0.58477056 1109 132872 1 1.3953474 1135 135045 0 0.0 1004 132955 2 2.2419016 1097 132955 2 2.078217 1115 135082 2 0.20140053 1018 135082 2 0.10643565 1103 135104 1 0.8050781 1082 132609 0 0.0 1050 132609 0 0.0 1070 132609 1 1.1171359 1063 132609 1 1.538623 1126 135054 1 1.2721417 1008 132869 1 0.6903402 1015 132869 2 0.5125378 1131 132869 1 1.0112096 1085 135059 2 1.5353825 1137 135059 2 1.7157091 1002 135059 1 1.3568485 1004 132958 2 1.7654879 1089 135028 1 0.81338346 1132 135028 2 1.2805307 1059 135028 1 1.3594056 1081 135028 1 0.43034786 1071 135028 2 1.652563 1033 135028 2 1.2927337 1106 135028 2 1.1417092 1004 135028 2 1.6059607 1109 135028 2 1.604281 1029 132922 1 0.9190735 1086 132922 2 1.917752 1090 132922 2 2.1243176 1138 132922 2 1.2562399 1072 135016 1 0.0 1117 135016 2 0.0 1022 135033 1 0.9656447 1104 135033 1 0.74237585 1016 1350